core/base/src/main/java/alluxio/collections/ConcurrentIdentityHashMap.java

```diff
@@ -0,0 +1,383 @@
+    return new Set<K>() {
+      @Override
+      public int size() {
+        return set.size();
+      }
+
+      @Override
+      public boolean isEmpty() {
+        return set.isEmpty();
+      }
+
+      @Override
+      public boolean contains(Object o) {
+        return set.contains(new IdentityObject<>(o));
+      }
+
+      @Override
+      public Iterator<K> iterator() {
+        return set.stream().map(IdentityObject::get).iterator();
+      }
+
+      @Override
+      public Object[] toArray() {
+        throw new UnsupportedOperationException("toArray not supported in identity keySet");
+      }
+
+      @Override
+      public <T> T[] toArray(T[] a) {
+        throw new UnsupportedOperationException("toArray not supported in identity keySet");
+      }
+
+      @Override
+      public boolean add(K k) {
+        throw new UnsupportedOperationException("add not supported in keySet");
+      }
+
+      @Override
+      public boolean remove(Object o) {
+        return set.remove(new IdentityObject<>(o));
+      }
+
+      @Override
+      public boolean containsAll(Collection<?> c) {
+        return set.containsAll(c);
+      }
+
+      @Override
+      public boolean addAll(Collection<? extends K> c) {
+        throw new UnsupportedOperationException("addAll not supported in keySet");
+      }
+
+      @Override
+      public boolean retainAll(Collection<?> c) {
+        return set.retainAll(c);
+      }
+
+      @Override
+      public boolean removeAll(Collection<?> c) {
+        return set.retainAll(c);
+      }
+
+      @Override
+      public void clear() {
+        set.clear();
+      }
+    };
+  }
+
+  private class IdentityObject<T> {
+    private T mObj;
+
+    public IdentityObject(T o) {
+      mObj = o;
+    }
+
+    public T get() {
+      return mObj;
+    }
+
+    /*
+     * This method hashes on top of the {@link System#identityHashCode(Object)} because the
+     * {@link ConcurrentHashMap} implementation uses a bucketing strategy where hash collisions
+     * occur for when values do not differ in upper or lower bits. Using the extra hash here
+     * helps ensure more even spread among the buckets of the internal map
+     */
+    @Override
+    public int hashCode() {
+      // Robert Jenkins 32-bit integer hash
+      int a = System.identityHashCode(mObj);
+      a = (a + 0x7ed55d16) + (a << 12);
+      a = (a ^ 0xc761c23c) ^ (a >> 19);
+      a = (a + 0x165667b1) + (a << 5);
+      a = (a + 0xd3a2646c) ^ (a << 9);
+      a = (a + 0xfd7046c5) + (a << 3);
+      a = (a ^ 0xb55a4f09) ^ (a >> 16);
+      return a;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (!(o instanceof IdentityObject)) {
+        // If it isn't an instance of an identity object, compare our internal pointer with the arg
+        // A form of "auto unboxing" when performing comparisons
+        return mObj == o;
+      }
```

Comment:
This looks dangerous. It may hide some issues in the future if we change the implementation. Do we really need unboxing?

Reply:
Not necessary but I thought it was kind of neat when writing that I could write this such that I could compare an `IdentityObject<T>` with an instance of `T` to check if they were equal without having to do  `IdentityObject<T>.equals(new IdentityObject<T>(other))`

Reply:
I can change this to simply return false